src/commands.ts

```diff
@@ -323,16 +257,156 @@ export function editGroup(treeItem: vscode.TreeItem, state: State) {
+    state.groups.splice(deleteIndex, 1);
+  }
+  refreshEditors(state);
+}
+
+export function saveProject(state: State) {
+  if (state.groups.length === 0) {
+    vscode.window.showErrorMessage('There is no filter groups');
+    return;
+  }
+
+  const selected = state.projects.find(p => (p.selected === true));
+  if (selected === undefined) {
+    vscode.window.showErrorMessage('There is no selected project');
+    return;
+  }
+
+  selected.groups = state.groups;
+  saveSettings(state.projects);
+
+  setStatusBarMessage(`Project(${selected.name}) is saved.`);
+}
+
+export function addProject(state: State) {
+  vscode.window.showInputBox({
+    prompt: "[PROJECT] Type a new project name",
+    ignoreFocusOut: false
+  }).then(name => {
+    if (name === undefined) {
+      return;
+    }
+
+    const project = {
+      groups: [],
+      name,
+      id: `${Math.random()}`,
+      selected: false
+    };
+
+    state.projects.push(project);
+    saveSettings(state.projects);
+    updateProjectAndTreeView(state);
+  });
+}
+
+export function editProject(treeItem: vscode.TreeItem, state: State, callback: () => void) {
+  vscode.window
+    .showInputBox({
+      prompt: "[PROJECT] Type a new name",
+      ignoreFocusOut: false,
+    })
+    .then((name) => {
+      if (name === undefined) {
+        return;
+      }
+      const findIndex = state.projects.findIndex(project => (project.id === treeItem.id));
+      if (findIndex !== -1) {
+        state.projects[findIndex].name = name;
+        saveSettings(state.projects);
+        updateProjectAndTreeView(state);
+
+        callback();
+      }
+    });
+}
+
+export function deleteProject(treeItem: vscode.TreeItem, state: State) {
+  const deleteIndex = state.projects.findIndex(project => (project.id === treeItem.id));
   if (deleteIndex !== -1) {
-    state.groupArr.splice(deleteIndex, 1);
+    if (deleteIndex === state.selectedIndex) {
+      state.groups = [];
+      state.selectedIndex = -1;
+      updateFilterAndTreeView(state);
+      refreshEditors(state);
+    }
+    state.projects.splice(deleteIndex, 1);
+    saveSettings(state.projects);
+    updateProjectAndTreeView(state);
+  }
+}
+
+export function refreshSettings(state: State) {
+  state.projects = readSettings();
+
+  // Automatically activate the project if there is only one
+  if (state.projects.length === 1) {
+    state.selectedIndex = 0;
   }
+
+  // Add a project named "NONAMED" in the following cases:
+  // - A default project is generated for users who do not use the project feature.
+  // - If multiple projects are available but none is selected, an empty project is created and selected.
+  if (state.projects.length === 0 || state.selectedIndex === -1) {
+    const name = "NONAME";
+    const project = {
+      groups: [],
+      name,
+      id: `${Math.random()}`,
+      selected: false
+    };
+
+    const projectLength = state.projects.push(project);
+    state.selectedIndex = projectLength - 1;
+  }
```

Comment:
There is no duplication because if there are no projects, a `NONAME` project is created by reading the configuration file. I have modified it so that after creating the `NONAME` project, it is saved to the file.

Reply:
Saw the new commit, fine with me.